### PR TITLE
ng_net: changed semantic of get and set command slightly

### PR DESCRIPTION
**Pull request by haukepetersen**
When working with netdev/netapi, I found the get function over complicated. This PR simplifies the semantics a little by making use of the `netdev->get()` functions return value (like every Linux programmer would expect)...

So when reading for example a network address from a device, the max_length parameter can be used to specify how many bytes should be read. The device driver can then take from this number which address it should read, e.g. a 802.15.4 device returns the short address if 2 byte are asked for and the long address if 8 byte are asked for...

### CODE REVIEW

File: sys/shell/commands/sc_netif.c [hunk 1]
```diff
@@ -120,15 +120,15 @@ void _netif_send(int argc, char **argv)
...
 {
     kernel_pid_t dev;
     uint8_t addr[MAX_ADDR_LEN];
     size_t addr_len;
     int res;
 
-    if (argc < 2) {
-        printf("usage: %s IF [ADDR]\n", argv[0]);
+    if (argc < 3) {
+        printf("usage: %s IF ADDR_LEN (in byte)\n", argv[0]);
```
Comment (authmillenon) [inline sys/shell/commands/sc_netif.c]: Where does the user get this information? Why do we need it in addition to `MAX_ADDR_LEN`?

**Comment by authmillenon** [comment]
We also should note, that most drivers will error if `max_len != sizeof(type)` for scalar values. Otherwise the implementation will get quite hairy ;-).

**Comment by authmillenon** [comment]
Apart from the split of the shell command, which seems a bit nonsensical to me :+1:

**Comment by authmillenon** [comment]
(it also would collide with #2581 ;-P)

File: sys/shell/commands/sc_netif.c (hunk 1 above)
Comment (haukepetersen) [inline sys/shell/commands/sc_netif.c]: This parameter is basically used to distinguish between different address types (short/long). If the user does not specify any length, there would be no way in knowing which address the user wants to know...

**Comment by haukepetersen** [comment]
The split is necessary, as both `addr_set` and `addr_get` have 3 parameters. Alternatively I could introduce a forth parameter, but this seemed easier to me (as #2581 does change this again anyway?!)

saw #2581 only after I did these changes...

**Comment by authmillenon** [comment]
> The split is necessary, as both addr_set and addr_get have 3 parameters. Alternatively I could introduce a forth parameter, but this seemed easier to me (as #2581 does change this again anyway?!)

I don't get why you need 3 parameters for `addr_get`. We have `MAX_ADDR_LEN` to determine `max_len`. Why does the user has to type in an address length manually if there is still a check directly afterwards if it is `> MAX_ADDR_LEN` (which yields an error).

File: sys/shell/commands/sc_netif.c [hunk 2]
```diff
@@ -137,37 +137,52 @@ void _netif_addr(int argc, char **argv)
...
+        puts("error: invalid interface given");
+        return;
+    }
+    addr_len = _parse_addr(argv[2], addr);
+    if (addr_len == 0) {
+        puts("error: unable to parse address");
+        return;
+    }
+    res = ng_netapi_set(dev, NETCONF_OPT_ADDRESS, 0, addr, addr_len);
+    if (res == addr_len) {
```
Comment (authmillenon) [inline sys/shell/commands/sc_netif.c]: So is `addr_len` the maximum length now or the exact length? I think your change to `set()`, though understandable to keep it same to `get()`, is unclear here.

Comment (authmillenon) [inline sys/shell/commands/sc_netif.c]: Okay… I guess I have to be clearer here: What I normally do with array-like values like addresses: I call `ng_netapi_set(dev, NETCONF_OPT_ADDRSS, 0, addr, sizeof(addr))`, where `addr` is an array of arbitrary length. Would this be an error case? If yes, why?

Comment (haukepetersen) [inline sys/shell/commands/sc_netif.c]: The meaning of the `len` parameter is quite straight forward in my opinion. When calling `set`, you specify the length of the given address (as you state in your example above). For interfaces, that support addresses of different length, this parameter can be used to detect which address should be set (short vs. long address). So it is possible to set both addresses with the same function.

When calling `get`, the `max_len` parameter is used in the same way. If I would just give the `MAX_ADDR_LEN`, the driver would have no means of finding out, which address is actually requested. By setting `max_len` to say 2 or 8, the driver can read from this context, which address the user wants to read.

So for `set`, the `len` parameter defines the exact length of the data to be written to the device. For `get`, the `max_len` parameter defines for one the maximum number of bytes that can be written to the data buffer, but gives also a context for the amount of data that is expected.

Clearer now?

**Comment by haukepetersen** [comment]
The third parameter for `addr_get` is important to specify the expected length of the address that is read - this is important to differentiate between different device address (short vs. long address). Without this parameter, there would be no means of telling the driver, which address you want to get.

The `MAX_ADDR_LEN` macro is simply there to define the internal array size that is used to save the address internally - it's value is chosen arbitrarily...

**Comment by haukepetersen** [comment]
After re-consideration, it was decided that the idea of a unified shell command might not be as good as I thought. So I took the changes to the netif shell commands out of this PR...

**Comment by haukepetersen** [comment]
The reason is that is a times difficult to know the exact addr length, when using the get command in an automated environment (as for example the `ifconfig` shell command)...

**Comment by authmillenon** [comment]
ACK. Waiting for travis to be happy :-)

**Comment by haukepetersen** [comment]
Travis is happy. And go.

**Pull request merged**